base/compiler/typeinfer.jl

```diff
@@ -8,7 +8,179 @@ function typeinf(interp::AbstractInterpreter, result::InferenceResult, cached::B
+MethodInstance.
+"""
+struct Timing
+    mi_info::Any
+    start_time::UInt64
+    cur_start_time::UInt64
+    time::UInt64
+    children::Core.Array{Timing,1}
+    Timing(mi_info, start_time, cur_start_time, time, children) = new(mi_info, start_time, cur_start_time, time, children)
+    Timing(mi_info, start_time) = new(mi_info, start_time, start_time, UInt64(0), Timing[])
+end
+
+_time_ns() = ccall(:jl_hrtime, UInt64, ())  # Re-implemented here because Base not yet available.
+
+# We keep a stack of the Timings for each of the MethodInstances currently being timed.
+# Since type inference currently operates via a depth-first search (during abstract
+# evaluation), this vector operates like a call stack. The last node in _timings is the
+# node currently being inferred, and its parent is directly before it, etc.
+# Each Timing also contains its own vector for all of its children, so that the tree
+# call structure through type inference is recorded. (It's recorded as a tree, not a graph,
+# because we create a new node for duplicates.)
+const _timings = Timing[]
+"""
+    Core.Compiler.reset_timings()
+Empty out the previously recorded type inference timings, and start the "root" timer again.
+"""
+function reset_timings()
+    Core.Compiler.empty!(_timings)
+    Core.Compiler.push!(_timings, Timing("root", _time_ns()))
+    nothing
+end
+reset_timings()
+
+# (This is split into a function so that it can be called both here and once at the Very
+# End of the operation, by whoever started the operation and called `reset_timings()`.)
+@inline function close_current_timer()
+    stop_time = _time_ns()
+    parent_timer = _timings[end]
+    accum_time = Core.Compiler.:(-)(stop_time, parent_timer.cur_start_time)
+
+    # Add in accum_time ("modify" the immutable struct)
+    @inbounds begin
+        _timings[end] = Timings.Timing(
+            parent_timer.mi_info,
+            parent_timer.start_time,
+            parent_timer.cur_start_time,
+            parent_timer.time + accum_time,
+            parent_timer.children,
+        )
+    end
+    return nothing
+end
+
+@inline function enter_new_timer(frame)
+    # Very first thing, stop the active timer: get the current time and add in the
+    # time since it was last started to its aggregate exclussive time.
+    close_current_timer()
+
+    mi_info = _typeinf_identifier(frame)
+
+    # Start the new timer right before returning
+    push!(_timings, Timings.Timing(mi_info, UInt64(0)))
+    len = length(_timings)
+    new_timer = @inbounds _timings[len]
+    # Set the current time _after_ appending the node, to try to exclude the
+    # overhead from measurement.
+    start = Timings._time_ns()
+
+    @inbounds begin
+        _timings[len] = Timings.Timing(
+            new_timer.mi_info,
+            start,
+            start,
+            new_timer.time,
+            new_timer.children,
+        )
+    end
+
+    return nothing
+end
+
+@inline function exit_current_timer(_expected_frame_)
```

Comment:
Any reason not to `_expected_frame_` -> `frame`? The name seems to be signaling something but I'm not getting it.

Reply:
Thanks, good point. I think what I was trying trying to signal was that `frame` is not _actually_ needed at this step of the algorithm, i'm literally only passing it here for the error checking in the `@assert` to make sure that our assumptions hold (the assumption about DFS you noted in your previous review).

I could probably make that more explicit with a comment? Do you have any preferences? :) thanks for reviewing!

Reply:
A comment would be fine, but not a big deal either way.